docs/en/04_Changelogs/4.0.0.md

```diff
@@ -92,6 +93,138 @@ For a full list of renamed classes, check the `.upgrade.yml` definitions in each
+}
+```
+
+##### Class name remapping
+
+If you've namespaced one of your custom page types, you may notice a message in the CMS 
+ telling you it's obsolete. This is likely because the `ClassName` 
+field in the `SiteTree` table still contains the singular model name, e.g. `BlogPost`
+and that when you change it to `SilverStripe\Blog\Model\BlogPost` then everything 
+works again.
+
+Luckily the `dev/build` task is configured to look for a legacy class name mapping 
+configuration setting and will update this for you automatically. For an example 
+take a look at `_config/legacy.yml` in the CMS module. The Blog module, for instance, uses 
+this:
+
+```yaml
+SilverStripe\ORM\DatabaseAdmin:
+  classname_value_remapping:
+    Blog: SilverStripe\Blog\Model\Blog
+    BlogCategory: SilverStripe\Blog\Model\BlogCategory
+    BlogPost: SilverStripe\Blog\Model\BlogPost
+    BlogTag: SilverStripe\Blog\Model\BlogTag
+```
+
+##### Other namespacing gotchas
+
+You'll need to update any strings that represent class names and make sure they're fully
+qualified. In particular, relationship definitions such as `has_one` and `has_many` will need
+to be updated to refer to fully qualified class names.
+
+```ss
+<!-- before -->
+private static $has_one = [
+  'MyRelation' => 'MyObject',
+];
+<!-- after -->
+private static $has_one = [
+  'MyRelation' => MyObject::class,
+];
+```
+ 
+In the context of YAML, the magic constant `::class` does not apply. Class names must be hard coded.
+
+```ss
+<!-- before -->
+MyObject:
+  property: value
+<!-- after -->
+My\Project\MyObject:
+  property: value
+```
+
+When dealing with custom database queries, you need to be judicious about replacing table
+names expressed literally. The upgrader will automatically replace anything that looks
+like an unqualified class name with a fully qualified one, but a fully qualified
+class name may not be the same as the table name, as explained above, in the case of
+`$table_name`.
+
+For example:
+```php
+->innerJoin('BlogPost', sprintf('"BlogPost%s"."ID" = "SiteTree%s"."ID"', $stage, $stage))
+```
+The upgrader will replace `BlogPost` with `SilverStripe\\Blog\\Model\BlogPost` as the fully qualified class name. In the case of innerJoin(), it asks for a table name which have remained the same as SilverStripe 3.x, so this should not be namespaced. This of course depends on what you’ve configured your table name to be.
+
+A better option is to replace it with `DataObject::getSchema()->tableName(BlogPost::class)` as long as the class has already been imported into your current namespace.
+
+
+
+
+#### Migrate your controllers to their own files
+
+The convention for naming controllers is now `[MyPageType]Controller`, where it used to be `[MyPageType]_Controller`. This change was made to be more compatible with the PSR-2 standards.
+
+You can still use, for example, `Page_Controller`, but you will get a deprecation notice. It is best to change it to `PageController` during your upgrade process.
```

Comment:
This is true, but any module depending on the page controller will look for `PageController`